src/util/small_map.h

```diff
@@ -0,0 +1,560 @@
+/*******************************************************************\
+
+Module: Small map
+
+Author: Daniel Poetzl
+
+\*******************************************************************/
+
+/// \file
+/// Small map
+
+#ifndef CPROVER_UTIL_SMALL_MAP_H
+#define CPROVER_UTIL_SMALL_MAP_H
+
+#include <bitset>
+#include <cstdint>
+#include <cstring>
+#include <limits>
+#include <memory>
+#include <new>
+#include <type_traits>
+#include <utility>
+
+#include "invariant.h"
+
+// When we don't have constexpr
+#if !(defined(__GNUC__) || _MSC_VER >= 1900)
+
+template <std::size_t N>
+struct num_bitst
+{
+  static const std::size_t value = 1 + num_bitst<(N >> 1)>::value;
+};
+
+template <>
+struct num_bitst<1>
+{
+  static const std::size_t value = 1;
+};
+
+template <typename T, std::size_t B, typename U = std::integral_constant<T, 1>>
+struct indicator_maskt
+{
+  static const T value =
+    U::value |
+    indicator_maskt<T, B, std::integral_constant<T, (U::value << B)>>::value;
+};
+
+template <typename T, std::size_t B>
+struct indicator_maskt<T, B, std::integral_constant<T, 0>>
+{
+  static const T value = 0;
+};
+
+#endif
+
+/// Map from small integers to values
+///
+/// A data structure that maps small integers (in {0, ..., Num-1}) to values.
+/// It is designed to be more memory-efficient than std::map for this use case.
+///
+/// In the following we give some data about the memory consumption of
+/// small_mapt compared to std::map, on a 64-bit Linux system with GNU STL. It
+/// was determined with valgrind/massif. We configured the maps as
+/// `small_mapt<uint64_t, uint32_t, Num=8>` and `std::map<uint8_t, uint64_t>`.
+/// Thus, indices to small_mapt can be in {0, ..., 7} for this configuration.
+///
+/// Below we give the memory consumption (in bytes) of small_mapt and std::map,
+/// both for when containing 8 elements, and when being empty. The value for
+/// "useful" indicates the number of bytes requested, the value for "total" also
+/// includes padding bytes.
+///
+/// small_mapt:
+/// - 8 elements:
+///   * useful:  80 B
+///   * total:  120 B
+/// - empty:
+///   * useful:  16 B
+///   * total:   32 B
```

Comment:
Thank you for providing the data! What is the advantage of this implementation over one using a `forward_list` with a `std::map` style API put on top? (Which would be 8B when empty and probably 136B with 8 elements.)

Reply:
For some reason massif reports slightly higher values for the memory usage than what we would get by manually calculating it. Manually computing the memory usage of `small_mapt` we get 8B when empty, and 76B when containing 8 elements.

When using `std::forward_list`, the implementation of the map would be similar, except that when we currently read/modify the memory block we'd modify the list instead. We would still need a field to store the mapping from keys (e.g., integers in the range 0-7) to the index of the node in the list that stores the value (which would be in the range 0-3 for a map containing 4 elements). Including this field the map would use 140B when containing 8 elements.

Looking up an element in the current `small_mapt` is O(1), using `std::forward_list` it would be O(n). Removal and insertion are O(n) for both, though likely with a higher constant for the current `small_mapt` as it might require copying/moving elements.

Reply:
Also when using `std::forward_list` we wouldn't need key comparisons actually. That's because we have the member field `ind` (which is of type `uint32_t` by default) which stores the mapping from keys to the positions of the elements in the memory block/list. The methods `get_field()`/`set_field()` index into that field via a bitshift and masking operation.